crates/matrix-sdk-indexeddb/src/crypto_store/migrations/mod.rs

```diff
@@ -62,6 +62,40 @@ impl Drop for MigrationDb {
     }
 }
 
+/// The latest version of the schema we can support. If we encounter a database
+/// version with a higher schema version, we will return an error.
+///
+/// A note on how this works.
+///
+/// Normally, when you open an indexeddb database, you tell it the "schema
+/// version" that you know about. If the existing database is older than
+/// that, it lets you run a migration. If the existing database is newer, then
+/// it assumes that there have been incompatible schema changes and complains
+/// with an error ("The requested version (10) is less than the existing version
+/// (11)").
+///
+/// The problem with this is that, if someone upgrades their installed
+/// application, then realises it was a terrible mistake and tries to roll
+/// back, then suddenly every user's session is completely hosed. (They see
+/// an "unable to restore session" dialog.) Often, schema updates aren't
+/// actually backwards-incompatible — for example, existing code will work just
+/// fine if someone adds a new store or a new index — so this approach is too
+/// heavy-handed.
+///
+/// The solution we take here is to say "any schema changes up to
+/// [`MAX_SUPPORTED_SCHEMA_VERSION`] will be backwards-compatible". If, at some
+/// point, we do make a breaking change, we will give that schema version a
+/// higher number. Then, rather than using the implicit version check that comes
+/// with `indexedDB.open(name, version)`, we explicitly check the version
+/// ourselves.
+///
+/// It is expected that we will use version numbers that are multiples of 10 to
```

Comment:
I appreciate the proposal. Thanks for working on this!

I wonder if 10 versions before a breaking change is enough. May 100 be a safer interval? It doesn't hurt, but it gives more rooms for backward-compatible new versions.

What do you think?

Reply:
My main thinking here was that, if we find ourselves getting close to 20 without a breaking change, we can always bump `MAX_SUPPORTED_SCHEMA_VERSION` to 29, provided we do so before we actually *hit* 20.

You might be right though. Version numbers are cheap (we've got 4 billion of them to play with...)

Reply:
I've made this change in [41dfa88](https://github.com/matrix-org/matrix-rust-sdk/pull/3812/commits/41dfa888ac742bb264ab4186036bc676c58bd1b1)